File: 53-maximum-subarray/53-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    const n = nums.length
    
    for (let i = 1; i < n; i++) {
        nums[i] = Math.max(nums[i], nums[i] + nums[i - 1])
    }
    return Math.max(...nums)
};

// nums = [integers]
// if max is neg int
// FOR num in nums
// reduce function 
// if > curr
// if > max

// 

File: 53-maximum-subarray/53-maximum-subarray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        maxSub = nums[0]
        cur = 0
        
        for n in nums:
            if cur < 0:
                cur = 0
            cur += n
            maxSub = max(maxSub, cur)
        return maxSub